rskj-core/src/main/java/co/rsk/peg/federation/FederationSupportImpl.java

```diff
@@ -664,7 +680,77 @@ protected Integer commitFederation(boolean dryRun, Keccak256 hash, BridgeEventLo
+        return FederationChangeResponseCode.SUCCESSFUL;
+    }
+
+    /**
+     * Commits the currently pending federation to be the proposed one.
+     * That is, the proposed federation is set to be a new federation generated from the currently pending federation,
+     * and the pending federation is wiped out.
+     * @param dryRun whether to just do a dry run
+     * @param hash the pending federation's hash. This is checked to match the execution block's pending federation hash.
+     * @return SUCCESSFUL upon success, PENDING_FEDERATION_NON_EXISTENT if there was no pending federation,
+     * INSUFFICIENT_MEMBERS if the pending federation was incomplete,
+     * PENDING_FEDERATION_MISMATCHED_HASH if the given hash doesn't match the current pending federation's hash.
+     */
+    private FederationChangeResponseCode commitFederation(boolean dryRun, Keccak256 hash, BridgeEventLogger eventLogger) {
+        PendingFederation currentPendingFederation = provider.getPendingFederation();
+
+        if (currentPendingFederation == null) {
+            logger.warn("[commitFederation] Pending federation does not exist.");
+            return FederationChangeResponseCode.PENDING_FEDERATION_NON_EXISTENT;
+        }
+
+        if (!currentPendingFederation.isComplete()) {
+            logger.warn("[commitFederation] Pending federation has {} members, so it does not meet the minimum required.", currentPendingFederation.getMembers().size());
+            return FederationChangeResponseCode.INSUFFICIENT_MEMBERS;
+        }
+
+        if (!hash.equals(currentPendingFederation.getHash())) {
+            logger.warn("[commitFederation] Provided hash {} does not match pending federation hash {}.", hash, currentPendingFederation.getHash());
+            return FederationChangeResponseCode.PENDING_FEDERATION_MISMATCHED_HASH;
+        }
+
+        if (dryRun) {
+            logger.info("[commitFederation] DryRun execution successful.");
+            return FederationChangeResponseCode.SUCCESSFUL;
+        }
+
+        Instant creationTime = Instant.ofEpochMilli(rskExecutionBlock.getTimestamp());
+        long rskExecutionBlockNumber = rskExecutionBlock.getNumber();
+
+        Federation proposedFederation =
+            currentPendingFederation.buildFederation(creationTime, rskExecutionBlockNumber, constants, activations);
+        provider.setProposedFederation(proposedFederation);
+
+        // Clear pending federation and votes on election
+        provider.clearPendingFederation();
+        provider.getFederationElection(constants.getFederationChangeAuthorizer()).clear();
+
+        Federation activeFederation = getActiveFederation();
+        eventLogger.logCommitFederation(rskExecutionBlock, activeFederation, proposedFederation);
+        logger.debug("[commitProposedFederation] Proposed Federation committed: {}", proposedFederation.getAddress());
+
+        if (activations.isActive(RSKIP186)) {
```

Comment:
Shouldn't we assume this is active at this point?

Reply:
good one, thanks

Reply:
done